### DISCUSSION TITLE
excluding some characters from string

### DISCUSSION THREAD

**Discussion by MahaZaid** [taocpp/PEGTL#201] [Q&A] [answered]
Hi, 
I want to match a keyword that does not include some specific words or characters, something like using regular expression for the matched keywords. 
for example I want to match any sequence of characters that generates a word except `AND` keyword
any idea how I could do that using PEGTL?

**Comment by d-frey** [comment, marked as answer]
Start by having rules for your keywords:

    struct kw_and : TAO_PEGTL_KEYWORD("AND") {};
    struct kw_or : TAO_PEGTL_KEYWORD("OR") {};

Then have a rule for *any* keyword:

    struct keywords : sor< kw_and, kw_or > {};

Finally, use the `minus<>` rule to match an identifier and make sure it is *not* a keyword.

    struct label : minus< identifier, keywords > {};

This should do the trick. You can, of course, use more complicated grammars if you want to exclude everything that contains a keyword or more complicated patterns.

**Comment by MahaZaid** [comment]
okay cool, but is identifier by default includes `space` characters ? as when I try to match 
`test1   test2` it passes, while I want it to fail in my case.
I thought identifier supports only letters & numbers.
here is my grammar:

```
   struct kw_and : TAO_PEGTL_STRING("AND") {};
    struct kw_or : TAO_PEGTL_STRING("OR") {};
    struct keywords : sor<kw_and, kw_or> {};
    struct label : minus< identifier, keywords> {};
    struct grammar: must<label>{};
```

**Comment by d-frey** [comment]
That's just because the grammar matches the "test1" part of the input and ignores the rest. You need to explicitly enforce that the input ends after your label. Use:

    struct grammar : must< label, eof > {};

**Comment by MahaZaid** [comment]
It worked! 
thank you so much for your help :)

**Comment by d-frey** [comment]
You're welcome, thanks for choosing the PEGTL :)
